subxt/src/backend/unstable/follow_stream_driver.rs

```diff
@@ -0,0 +1,488 @@
+            // and return None to signal done-ness.
+            let Some(items) = items else {
+                self.done = true;
+                return Poll::Ready(None);
+            };
+
+            // No items? We've saved the waker so we'll be told when more come.
+            // Else, save the items locally and loop around to pop from them.
+            if items.is_empty() {
+                return Poll::Pending;
+            } else {
+                self.local_items = items;
+            }
+        }
+    }
+}
+
+impl<Hash: BlockHash> FollowStreamDriverSubscription<Hash> {
+    /// Return the current subscription ID. If the subscription has stopped, then this will
+    /// wait until a new subscription has started with a new ID.
+    pub async fn subscription_id(self) -> Option<String> {
+        let ready_event = self
+            .skip_while(|ev| std::future::ready(!matches!(ev, FollowStreamMsg::Ready(_))))
+            .next()
+            .await?;
+
+        match ready_event {
+            FollowStreamMsg::Ready(sub_id) => Some(sub_id),
+            _ => None,
+        }
+    }
+
+    /// Subscribe to the follow events, ignoring any other messages.
+    pub fn events(self) -> impl Stream<Item = FollowEvent<BlockRef<Hash>>> + Send + Sync {
+        self.filter_map(|ev| std::future::ready(ev.into_event()))
+    }
+}
+
+impl<Hash: BlockHash> Clone for FollowStreamDriverSubscription<Hash> {
+    fn clone(&self) -> Self {
+        self.shared.subscribe()
+    }
+}
+
+impl<Hash: BlockHash> Drop for FollowStreamDriverSubscription<Hash> {
+    fn drop(&mut self) {
+        self.shared.remove_sub(self.id);
+    }
+}
+
+/// Locked shared state.
+#[derive(Debug, Clone)]
+struct Shared<Hash: BlockHash>(Arc<Mutex<SharedState<Hash>>>);
+
+#[derive(Debug)]
+struct SharedState<Hash: BlockHash> {
+    done: bool,
+    next_id: usize,
+    subscribers: HashMap<usize, SubscriberDetails<Hash>>,
+    // Keep a buffer of all events from last finalized block so that new
+    // subscriptions can be handed this info first.
+    block_events_from_last_finalized: VecDeque<Ev<Hash>>,
+    // Keep track of the subscription ID we send out on new subs.
+    current_subscription_id: Option<String>,
+    // Keep track of the init message we send out on new subs.
+    current_init_message: Option<Initialized<BlockRef<Hash>>>,
+    // Runtime events by block hash; we need to track these to know
+    // whether the runtime has changed when we see a finalized block notification.
+    seen_runtime_events: HashMap<Hash, RuntimeEvent>,
+}
+
+impl<Hash: BlockHash> Default for Shared<Hash> {
+    fn default() -> Self {
+        Shared(Arc::new(Mutex::new(SharedState {
+            next_id: 1,
+            done: false,
+            subscribers: HashMap::new(),
+            current_init_message: None,
+            current_subscription_id: None,
+            seen_runtime_events: HashMap::new(),
+            block_events_from_last_finalized: VecDeque::new(),
```

Comment:
Nit: Why do we set `next_id: 1` instead of having it initialized to 0? 
I think we can also derive the `Default` trait for `SharedState`, to use `SharedState { next_id: 1, ..Default::default() }` or `Default::default()` here.

Reply:
For no particular reason other than to avoid `0` being used, which sometimes could be seen as a special or meaningful value. But there would be no harm in this being initialized to 0 either really; just a habit I adopted a bit!